File: src/main/java/gargoyle/xenox/util/applet/GFrame.java
package gargoyle.xenox.util.applet;

import gargoyle.xenox.util.log.Log;
import gargoyle.xenox.util.res.Resources;
import gargoyle.xenox.util.res.audio.AudioClip;

import javax.swing.BorderFactory;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.SwingUtilities;
import javax.swing.WindowConstants;
import javax.swing.border.BevelBorder;

import java.awt.BorderLayout;
import java.awt.Desktop;
import java.awt.Dimension;
import java.awt.GraphicsEnvironment;
import java.awt.Image;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.event.WindowEvent;
import java.awt.event.WindowListener;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.Collections;
import java.util.Enumeration;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

final class GFrame extends JFrame implements WindowListener, AppletStub, AppletContext {
    private static final String STR_READY = "Ready";
    private final GApplet applet;
    private final Map<String, String> parameters = new HashMap<>();
    private final JLabel status;
    private final Map<String, InputStream> streams = new HashMap<>();
    private final Resources resources;

    public GFrame(GApplet applet, String[] args, Resources resources) {
        this.applet = applet;
        this.resources = resources;
        applet.application = this;
        applet.setStub(this);
        setArgs(args);
        setTitle(applet.getTitle());
        setLayout(new BorderLayout());
        add(applet, BorderLayout.CENTER);
        status = new JLabel(STR_READY);
        status.setBorder(BorderFactory.createBevelBorder(BevelBorder.LOWERED));
        add(status, BorderLayout.SOUTH);
        setDefaultCloseOperation(WindowConstants.DO_NOTHING_ON_CLOSE);
        addWindowListener(this);
        Dimension
                size =
                new Dimension(GraphicsEnvironment.getLocalGraphicsEnvironment().getMaximumWindowBounds().getSize());
        size.width >>= 1;
        size.height >>= 1;
        applet.setPreferredSize(size);
        pack();
        setLocationRelativeTo(null);
    }

    void exit() {
        applet.destroy();
        SwingUtilities.invokeLater(() -> {
            setVisible(false);
            dispose();
        });
    }

    @Override
    public AudioClip getAudioClip(URL url) {
        return resources.load(AudioClip.class, url);
    }

    @Override
    public Image getImage(URL url) {
        return resources.load(BufferedImage.class, url);
    }

    @Override
    public Applet getApplet(String name) {
//        return applet;
        return null;
    }

    @Override
    public Enumeration<Applet> getApplets() {
//        return Collections.enumeration(Arrays.asList(new Applet[]{applet}));
        return Collections.emptyEnumeration();
    }

    @Override
    public URL getDocumentBase() {
        try {
            return new URL(new File(System.getProperty("user.dir", ".")).getCanonicalFile().toURI().toURL(), "");
        } catch (IOException e) {
            Log.error(e.getLocalizedMessage(), e);
            return null;
        }
    }

    @Override
    public URL getCodeBase() {
        try {
            return getClass().getProtectionDomain().getCodeSource().getLocation();
        } catch (RuntimeException e) {
            Log.error(e.getLocalizedMessage(), e);
            return null;
        }
    }

    @Override
    public void showStatus(String status) {
        if (isVisible()) {
            this.status.setText(status);
        }
    }

    @Override
    public void setStream(String key, InputStream stream) {
        streams.put(key, stream);
    }

    @Override
    public InputStream getStream(String key) {
        return streams.get(key);
    }

    @Override
    public Iterator<String> getStreamKeys() {
        return streams.keySet().iterator();
    }

    @Override
    public void showDocument(URL url) {
        try {
            if (Desktop.isDesktopSupported()) {
                Desktop.getDesktop().browse(url.toURI());
            }
        } catch (IOException | URISyntaxException e) {
            Log.error(e.getLocalizedMessage(), e);
        }
    }

    @Override
    public void showDocument(URL url, String target) {
        try {
            if (Desktop.isDesktopSupported()) {
                Desktop.getDesktop().browse(url.toURI());
            }
        } catch (IOException | URISyntaxException e) {
            Log.error(e.getLocalizedMessage(), e);
        }
    }

    @Override
    public String getParameter(String name) {
        return parameters.get(name);
    }

    @Override
    public AppletContext getAppletContext() {
        return this;
    }

    @Override
    public void appletResize(int width, int height) {
        synchronized (getTreeLock()) {
            Rectangle screen = GraphicsEnvironment.getLocalGraphicsEnvironment().getMaximumWindowBounds();
            Rectangle size = new Rectangle(getLocation(), applet.getMinimumSize());
            if (width > size.width) {
                size.width = width;
            }
            if (height > size.height) {
                size.height = height;
            }
            if (size.width > screen.width) {
                size.width = screen.width;
            }
            if (size.height > screen.height) {
                size.height = screen.height;
            }
            int x = screen.x + screen.width - size.width;
            if (size.x > x) {
                size.x = x;
            }
            int y = screen.y + screen.height - size.height;
            if (size.y > y) {
                size.y = y;
            }
            applet.setPreferredSize(size.getSize());
            applet.setSize(size.getSize());
            Point location = size.getLocation();
            Point c = new Point(getX() + getWidth() / 2, getY() + getHeight() / 2);
            setLocation(location);
            pack();
            Point c2 = new Point(getX() + getWidth() / 2, getY() + getHeight() / 2);
            setLocation(getX() - (c2.x - c.x), getY() - (c2.y - c.y));
        }
    }

    private void setArgs(String[] args) {
        for (String arg : args) {
            String[] split = arg.split("=", 2);
            parameters.put(split[0], split[1]);
        }
    }

    public void showMe() {
        SwingUtilities.invokeLater(() -> {
            setVisible(true);
            applet.init();
        });
    }

    @Override
    public void windowOpened(WindowEvent e) {
    }

    @Override
    public void windowClosing(WindowEvent e) {
        try {
            if (applet.canExit()) {
                applet.exit();
            } else {
                if (!isVisible()) {
                    setVisible(true);
                }
            }
        } catch (RuntimeException e1) {
            applet.error(e1);
            dispose();
        }
    }

    @Override
    public void windowClosed(WindowEvent e) {
    }

    @Override
    public void windowIconified(WindowEvent e) {
    }

    @Override
    public void windowDeiconified(WindowEvent e) {
    }

    @Override
    public void windowActivated(WindowEvent e) {
        applet.start();
    }

    @Override
    public void windowDeactivated(WindowEvent e) {
        applet.stop();
    }
}

File: src/main/kotlin/gargoyle/xenox/util/applet/GFrame.kt
package gargoyle.xenox.util.applet

import gargoyle.xenox.util.log.Log
import gargoyle.xenox.util.res.Resources
import gargoyle.xenox.util.res.audio.AudioClip
import java.awt.BorderLayout
import java.awt.Desktop
import java.awt.Dimension
import java.awt.GraphicsEnvironment
import java.awt.Image
import java.awt.Point
import java.awt.Rectangle
import java.awt.event.WindowEvent
import java.awt.event.WindowListener
import java.awt.image.BufferedImage
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.net.URISyntaxException
import java.net.URL
import java.util.Collections
import java.util.Enumeration
import javax.swing.BorderFactory
import javax.swing.JFrame
import javax.swing.JLabel
import javax.swing.SwingUtilities
import javax.swing.border.BevelBorder

class GFrame(private val applet: GApplet, args: Array<String>, private val resources: Resources) : JFrame(),
    WindowListener, AppletStub, AppletContext {
    private val parameters: MutableMap<String, String> = HashMap()
    private val status: JLabel
    private val streams: MutableMap<String, InputStream> = HashMap()

    init {
        applet.application = this
        applet.setStub(this)
        setArgs(args)
        title = applet.title
        layout = BorderLayout()
        add(applet, BorderLayout.CENTER)
        status = JLabel(STR_READY)
        status.border = BorderFactory.createBevelBorder(BevelBorder.LOWERED)
        add(status, BorderLayout.SOUTH)
        defaultCloseOperation = DO_NOTHING_ON_CLOSE
        addWindowListener(this)
        val size = Dimension(GraphicsEnvironment.getLocalGraphicsEnvironment().maximumWindowBounds.size)
        size.width = size.width shr 1
        size.height = size.height shr 1
        applet.preferredSize = size
        pack()
        setLocationRelativeTo(null)
    }

    fun exit() {
        applet.destroy()
        SwingUtilities.invokeLater {
            isVisible = false
            dispose()
        }
    }

    override fun getAudioClip(url: URL): AudioClip {
        return resources.load(AudioClip::class.java, url)!!
    }

    override fun getImage(url: URL): Image {
        return resources.load(BufferedImage::class.java, url)!!
    }

    override fun getApplet(name: String): Applet? {
//        return applet;
        return null
    }

    //        return Collections.enumeration(Arrays.asList(new Applet[]{applet}));
    override val applets: Enumeration<Applet>
        get() =//        return Collections.enumeration(Arrays.asList(new Applet[]{applet}));
            Collections.emptyEnumeration()
    override val documentBase: URL?
        get() = try {
            URL(File(System.getProperty("user.dir", ".")).canonicalFile.toURI().toURL(), "")
        } catch (e: IOException) {
            Log.error(e.localizedMessage, e)
            null
        }
    override val codeBase: URL?
        get() = try {
            javaClass.protectionDomain.codeSource.location
        } catch (e: RuntimeException) {
            Log.error(e.localizedMessage, e)
            null
        }

    override fun showStatus(status: String) {
        if (isVisible) {
            this.status.text = status
        }
    }

    override fun setStream(key: String, stream: InputStream) {
        streams[key] = stream
    }

    override fun getStream(key: String): InputStream? {
        return streams[key]
    }

    override val streamKeys: Iterator<String>
        get() = streams.keys.iterator()

    override fun showDocument(url: URL) {
        try {
            if (Desktop.isDesktopSupported()) {
                Desktop.getDesktop().browse(url.toURI())
            }
        } catch (e: IOException) {
            Log.error(e.localizedMessage, e)
        } catch (e: URISyntaxException) {
            Log.error(e.localizedMessage, e)
        }
    }

    override fun showDocument(url: URL, target: String) {
        try {
            if (Desktop.isDesktopSupported()) {
                Desktop.getDesktop().browse(url.toURI())
            }
        } catch (e: IOException) {
            Log.error(e.localizedMessage, e)
        } catch (e: URISyntaxException) {
            Log.error(e.localizedMessage, e)
        }
    }

    override fun getParameter(name: String): String? {
        return parameters[name]
    }

    override val appletContext: AppletContext
        get() = this

    override fun appletResize(width: Int, height: Int) {
        synchronized(treeLock) {
            val screen = GraphicsEnvironment.getLocalGraphicsEnvironment().maximumWindowBounds
            val size = Rectangle(location, applet.minimumSize)
            if (width > size.width) {
                size.width = width
            }
            if (height > size.height) {
                size.height = height
            }
            if (size.width > screen.width) {
                size.width = screen.width
            }
            if (size.height > screen.height) {
                size.height = screen.height
            }
            val x = screen.x + screen.width - size.width
            if (size.x > x) {
                size.x = x
            }
            val y = screen.y + screen.height - size.height
            if (size.y > y) {
                size.y = y
            }
            applet.preferredSize = size.size
            applet.size = size.size
            val location = size.location
            val c = Point(getX() + getWidth() / 2, getY() + getHeight() / 2)
            setLocation(location)
            pack()
            val c2 = Point(getX() + getWidth() / 2, getY() + getHeight() / 2)
            setLocation(getX() - (c2.x - c.x), getY() - (c2.y - c.y))
        }
    }

    private fun setArgs(args: Array<String>) {
        for (arg in args) {
            val split = arg.split('=').toTypedArray()
            parameters[split[0]] = split[1]
        }
    }

    fun showMe() {
        SwingUtilities.invokeLater {
            isVisible = true
            applet.init()
        }
    }

    override fun windowOpened(e: WindowEvent) {}
    override fun windowClosing(e: WindowEvent) {
        try {
            if (applet.canExit()) {
                applet.exit()
            } else {
                if (!isVisible) {
                    isVisible = true
                }
            }
        } catch (e1: RuntimeException) {
            applet.error(e1)
            dispose()
        }
    }

    override fun windowClosed(e: WindowEvent) {}
    override fun windowIconified(e: WindowEvent) {}
    override fun windowDeiconified(e: WindowEvent) {}
    override fun windowActivated(e: WindowEvent) {
        applet.start()
    }

    override fun windowDeactivated(e: WindowEvent) {
        applet.stop()
    }

    companion object {
        private const val STR_READY = "Ready"
    }
}
